### PR TITLE
Test infrastructure 2

### PR DESCRIPTION
**Pull request by dolik-rce**
This is an attempt to continue with the work on test infrastructure that @masatake started in #14. It tries to accomplish the same goals:
- Provide the way to write a unit test.
- Run the unit tests at GitHub Actions.

The main change from the original approach is that this implementation is not focused on checking the generated code directly. It rather checks if it can be compiled and if the compiled binary works as expected. Checking the code is of course possible as well (via custom tests, see tests/README.md), but not implemented at this moment.

Please check if this general approach is acceptable. If it is, I can start to write more tests, to eventually cover all (or almost all) of the packcc features.

### CODE REVIEW

File: tests/test.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+main() {
+    set -e
+
+    export TESTDIR="$(cd "$(dirname "$0")" && pwd)"
+    export ROOTDIR="$TESTDIR/.."
+
+    cd "$TESTDIR"
+    clean
+
+    "${CC:-cc}" -o packcc $ROOTDIR/src/packcc.c
```
Comment (masatake) [inline tests/test.sh:30]: How about using packcc already built.

```
for cc in $ROOTDIR/build/*/bin/debug/packcc; do
   ls "$cc" &> /dev/null && continue
   CC=$cc
done
```

File: tests/README.md [hunk 2]
```diff
@@ -2,39 +2,53 @@
...
-For running test cases, we assume you have `bats-core-1.2.1`.
+For running test cases, we assume you have `bats-core`.
 
-If you don't have, you can install it from its tarball:
+If you don't have it installed, you can do so using your package manager or from its tarball:
 ```
 $ curl -L -o bats-core-1.2.1.tar.gz https://github.com/bats-core/bats-core/archive/v1.2.1.tar.gz &&
   tar zxvf bats-core-1.2.1.tar.gz &&
   cd bats-core-1.2.1 &&
   ./install.sh /usr/local
```
Comment (masatake) [inline tests/README.md:12]:
```
sudo ./instal.sh /usr/local
````
is better.

File: tests/test.sh (hunk 1 above)
Comment (masatake) [inline tests/test.sh:30]: Like CC, how about introducing PACKCC environment variable for specifying a packcc executable for testing?
Unless PACKCC is not specified, we can make a packcc executable here.

**Comment by masatake** [comment]
Thank you very much for taking my incomplete work.

This is much better.

It will be nice if this test harness can be integrated to per compiler Makefile like
```
diff --git a/build/clang/Makefile b/build/clang/Makefile
index 9079002..5c7a313 100644
--- a/build/clang/Makefile
+++ b/build/clang/Makefile
@@ -49,3 +49,5 @@ $(TMPDIR_R)/examples/%.c $(TMPDIR_R)/examples/%.h: $(SRCDIR)/examples/%.peg $(BI
 
 clean:
        rm -f $(BINS) $(SRCS)
+
+include ../test.make
diff --git a/build/gcc/Makefile b/build/gcc/Makefile
index eb6e489..0ad08b2 100644
--- a/build/gcc/Makefile
+++ b/build/gcc/Makefile
@@ -49,3 +49,5 @@ $(TMPDIR_R)/examples/%.c $(TMPDIR_R)/examples/%.h: $(SRCDIR)/examples/%.peg $(BI
 
 clean:
        rm -f $(BINS) $(SRCS)
+
+include ../test.make
diff --git a/build/test.make b/build/test.make
new file mode 100644
index 0000000..eff5786
--- /dev/null
+++ b/build/test.make
@@ -0,0 +1,2 @@
+check:
+       ...run tests/test.sh with proper PACKCC...
```
`make check` makes the way to run test cases GitHub Actions simple.

File: .github/workflows/testing.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      CC: ${{ matrix.compiler }}
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: build packcc
+      run: |
+        (
+        cd  build/$CC
+        make
+        )
```
Comment (masatake) [inline .github/workflows/testing.yml:28]: It is nice if we can put `make check`

File: .github/workflows/testing.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    env:
+      CC: ${{ matrix.compiler }}
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: build packcc
+      run: |
+        (
+        cd  build/$CC
+        make
```
Comment (masatake) [inline .github/workflows/testing.yml:27]: It is nice if we can put `make check`

File: .github/workflows/testing.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        )
+    - name: install bats
+      run: |
+        curl -L -o bats-core-1.2.1.tar.gz https://github.com/bats-core/bats-core/archive/v1.2.1.tar.gz
+        tar zxvf bats-core-1.2.1.tar.gz
+        cd bats-core-1.2.1 &&
+        ./install.sh /usr/local
+    - name: run test cases
```
Comment (masatake) [inline .github/workflows/testing.yml:35]: If we have `make check` we can remove this.

File: tests/test.sh (hunk 1 above)
Comment (dolik-rce) [reply tests/test.sh:30]: Yes, good idea.

File: .github/workflows/testing.yml (hunk 4 above)
Comment (dolik-rce) [reply .github/workflows/testing.yml]: Done.

File: .github/workflows/testing.yml (hunk 5 above)
Comment (dolik-rce) [reply .github/workflows/testing.yml]: Done.

**Comment by masatake** [comment]
ping @arithy.

**Comment by arithy** [comment]
Thank you very much for this great work! It's very cool.

> Please check if this general approach is acceptable. If it is, I can start to write more tests, to eventually cover all (or almost all) of the packcc features.

I agree with this approach. I greatly appreciate your offer.

**Pull request merged**